### PR TITLE
DNN: reduce the memory used in convolution layer

### PR DESCRIPTION
**Pull request by zihaomu**
Related issue: https://github.com/opencv/opencv/issues/22825

Before this patch, every convolution 3x3s1 layer keeps twore-packed weight parameter one for general convolution and another for Winograd convolution.

This PR proposes to let the 3x3s1 convolution layer save only one re-packed weight parameter at a time.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Win32
```

### CODE REVIEW

**Comment by zihaomu** [comment]
Hi @alalek, can you check how much memory is reduced by this PR? Thx.

**Comment by alalek** [comment]
```
$ /usr/bin/time ./bin/opencv_test_dnn --test_threads=4 --gtest_filter=Test_Darknet.read_yolo_voc_stream
...
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Test_Darknet
[ RUN      ] Test_Darknet.read_yolo_voc_stream
[       OK ] Test_Darknet.read_yolo_voc_stream (1397 ms)
[----------] 1 test from Test_Darknet (1397 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1397 ms total)
[  PASSED  ] 1 test.
1.48user 0.98system 0:01.56elapsed 157%CPU (0avgtext+0avgdata 2144812maxresident)k
40848inputs+0outputs (70major+965499minor)pagefaults 0swaps
```

~2.1 GB

**Comment by zihaomu** [comment]
Thanks for the testing. Looks like we reduce some memory but not much 2336348 -> 2144812.

I just take a look at the detail of `read_yolo_voc_stream`. It consists of conv 3x3s1, which will be run in the Winograd branch.
Since our implementation of Winograd needs to extend every [3x3] kernel (or weight) to [8x8], that means we need about 8*8/(3*3) = 7.1 times the memory size than the general convolution implementation.

My solution for these specific test cases is: to disable the Winograd branch.

**Comment by vpisarev** [comment]
I suggest not to disable Winograd, but rather disable certain tests if they take too much memory. 2.1Gb of memory is nothing by today's standards. If we have a system that has little memory, users should not just use heavy models on such systems.

Secondly, I hope, at some point we will finally add FP16 compute path into DNN. In this case on ARM systems with FP16 arithmetics Winograd weights will take 2x less space, i.e. roughly 1.1Gb.

**Comment by zihaomu** [comment]
> I suggest not to disable Winograd, but rather disable certain tests if they take too much memory.

Hi @vpisarev, this is what I have done in this PR. I have disabled some high memory consumption test cases.

**Comment by alalek** [comment]
2 problems with YOLOv3 / YOLOv4 tests left: http://pullrequest.opencv.org/buildbot/builders/precommit_windows32/builds/100094

If you want to disable them, then add 2GB "skip" tags.

---

> 2.1Gb of memory is nothing by today's standards.

This is not true for smartphones and IOT devices. This is always a problem on 32-bit platforms.

Also limited memory bandwidth is an actual gap of modern multi-core processors/SoC, so we should to avoid 3-7 times exploding of the memory consumption.

**Comment by zihaomu** [comment]
Hi @alalek, I have added the "CV_TEST_TAG_MEMORY_2GB" for corresponding test cases. But cases were not skipped by Win32 CI as expected. Can you give me more details advice on how to skip these cases? Thx.

**Comment by alalek** [comment]
> I have added the "CV_TEST_TAG_MEMORY_2GB" for corresponding test cases.

Where?
There are no commits with changes in performance tests.

**Comment by zihaomu** [comment]
My fault, I add the tag to the accuracy test, instead of the performance test. Thanks for your reply.

**Comment by zihaomu** [comment]
Hi @alalek. The PR has passed Win32 CI. 
In addition, I found there are some environment errors in the archive CI of Linux x64.

File: modules/dnn/perf/perf_net.cpp [hunk 1]
```diff
@@ -198,6 +198,10 @@ PERF_TEST_P_(DNNTestNetwork, Inception_v2_SSD_TensorFlow)
 
 PERF_TEST_P_(DNNTestNetwork, YOLOv3)
 {
+    applyTestTag(
+            CV_TEST_TAG_VERYLONG,
```
Comment (alalek) [inline modules/dnn/perf/perf_net.cpp:202]: > `CV_TEST_TAG_VERYLONG`

Why do we need to add this to resolve out of memory issue?

Comment (zihaomu) [reply modules/dnn/perf/perf_net.cpp]: Fixed.

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Comment by JulienMaille** [comment]
I'm getting a crash with OPENCV backend and CPU target and I suspect this appeared with that PR.
I will revert it and confirm, do you guys see anything suspicious?

image

**Comment by JulienMaille** [comment]
@zihaomu my bad, reverting that commit did not fix the issue.
However forcing winograd to false, remove the problem, when was that enabled for non NEON architectures?
(I'm working on a clean project without Cmake cache)
I can invite you to a repository with a model that reproduce the problem

**Comment by JulienMaille** [comment]
@zihaomu where would that come from? Which input image size are you inferring on?
I'm running it on an Intel i9-10900X, win11 x64, openCV compiled from commit
> 103212f - Merge pull request #22940 from alalek:build_warnings_msvc - 10 December 2022
you can see my compile flags in the ps1 script from the repository I shared with you

**Comment by zihaomu** [comment]
@JulienMaille, the input shape is `[1,1,256,256]` whis is model expacted.
Test at M1 chip (ARM) and AMD 5600X (X86).

**Comment by JulienMaille** [comment]
Could you share a zip with compiled dlls so I can test on my Intel cpu?

**Comment by zihaomu** [comment]
I compile with msvc 2022.

**Comment by zihaomu** [comment]
I compile with msvc 2022.

**Pull request merged**